doc/function.md

```diff
@@ -1,5 +1,255 @@
+- `[in] data`: User-provided data context. This will be passed back into the
+function when invoked later.
+
+Returns instance of a `Function` object.
+
+### Call
+
+Calls a Javascript function from a native add-on.
+
+```cpp
+Value Call(const std::initializer_list<napi_value>& args) const;
+```
+
+- `[in] args`: Initializer list of JavaScript values as `napi_value` representing
+the arguments of the function.
+
+Returns a Value representing the JavaScript object returned by the function.
+
+### Call
+
+Calls a Javascript function from a native add-on.
+
+```cpp
+Value Call(const std::vector<napi_value>& args) const;
+```
+
+- `[in] args`: List of JavaScript values as `napi_value` representing the
+arguments of the function.
+
+Returns a Value representing the JavaScript object returned by the function.
+
+### Call
+
+Calls a Javascript function from a native add-on.
+
+```cpp
+Value Call(size_t argc, const napi_value* args) const;
+```
+
+- `[in] argc`: The number of the arguments passed to the function.
+- `[in] args`: Array of JavaScript values as `napi_value` representing the
+arguments of the function.
+
+Returns a Value representing the JavaScript object returned by the function.
+
+### Call
+
+Calls a Javascript function from a native add-on.
+
+```cpp
+Value Call(napi_value recv, const std::initializer_list<napi_value>& args) const;
+```
+
+- `[in] recv`: The `this` object passed to the called function.
+- `[in] args`: Initializer list of JavaScript values as `napi_value` representing
+the arguments of the function.
+
+Returns a Value representing the JavaScript object returned by the function.
+
+### Call
+
+Calls a Javascript function from a native add-on.
+
+```cpp
+Value Call(napi_value recv, const std::vector<napi_value>& args) const;
+```
+
+- `[in] recv`: The `this` object passed to the called function.
+- `[in] args`: List of JavaScript values as `napi_value` representing the
+arguments of the function.
+
+Returns a Value representing the JavaScript object returned by the function.
+
+### Call
+
+Calls a Javascript function from a native add-on.
+
+```cpp
+Value Call(napi_value recv, size_t argc, const napi_value* args) const;
+```
+
+- `[in] recv`: The `this` object passed to the called function.
+- `[in] argc`: The number of the arguments passed to the function.
+- `[in] args`: Array of JavaScript values as `napi_value` representing the
+arguments of the function.
+
+Returns a Value representing the JavaScript object returned by the function.
+
+### MakeCallback
+
+Calls a Javascript function from a native add-on after an asynchronous operation.
+
+```cpp
+Value MakeCallback(napi_value recv, const std::initializer_list<napi_value>& args) const;
+```
+
+- `[in] recv`: The `this` object passed to the called function.
+- `[in] args`: Initializer list of JavaScript values as `napi_value` representing
+the arguments of the function.
+
+Returns a Value representing the JavaScript object returned by the function.
```

Comment:
I think we need to have somewhere, either in the intro or in the description of the MakeCallback function where you should use Call versus MakeCallback.  You should use Call if there is already a JavaScript function on the stack (for example when running a native method called from JavaScript), and only use MakeCallback in special cases like running OnError, OnOk in AsyncWorker which don't have an existing JavaScript function on the stack.